### ISSUE TITLE
False positive on `no-loop-func`

### ISSUE THREAD

**Issue by isiahmeadows** [eslint/eslint#9527]
**Tell us about your environment**

* **ESLint Version:** Online
* **Node Version:** N/A
* **npm Version:** N/A

**What parser (default, Babel-ESLint, etc.) are you using?** Default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
exports.rules = {
    "no-loop-func": "error"
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

See [this demo snippet](https://eslint.org/demo/#eyJ0ZXh0IjoiXCJ1c2Ugc3RyaWN0XCI7XG5mdW5jdGlvbiBmb28oKSB7XG5cdGxldCBjcmVhdGUgPSAoKSA9PiB7fVxuXHRsZXQgdmFsdWVzID0gW11cblx0bGV0IHJlYWR5XG5cdFxuXHRmb3IgKGxldCBpID0gMDsgaSA8IDEwOyBpKyspIHtcblx0XHRjcmVhdGUoKCkgPT4ge1xuXHRcdFx0dmFsdWVzID0gcmVhZHlcblx0XHRcdHJlYWR5ID0gaVxuXHRcdH0pXG5cdH1cblxuXHRyZXR1cm4gKCkgPT4ge1xuXHRcdHJlYWR5ID0gMTBcblx0fVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6eyJpbXBsaWVkU3RyaWN0Ijp0cnVlfX0sInJ1bGVzIjp7Im5vLWxvb3AtZnVuYyI6Mn0sImVudiI6e319fQ==)

**What did you expect to happen?** No error

**What actually happened? Please include the actual, raw output from ESLint.**

It complained in the closure.

When you change `ready` to a loop-invariant variable (like `i`), the warning disappears. Note that in this reduced case (as well as my actual code), the closed-over variable being referenced outside the scope is provably loop-invariant outside the nested callback itself (and thus avoids the usual issues), even though it *is* modified in child scopes, and this *can* be discerned without the use of abstract evaluation. (Technically, abstract evaluation would just obliterate it entirely.)

-----

Note: it is out of the scope of this bug whether to allow code like this, where `ready` is *not* loop-invariant:

```js
"use strict";
function foo() {
	let create = () => {}
	let values = []
	let ready
	
	for (let i = 0; i < 10; i++) {
		create(() => {
			values = ready
		})
		ready = i
	}

	return () => {
		ready = 10
	}
}
```

**Comment by mysticatea** [comment]
Thank you for this issue.

However, this is working as intended.

Originally, the rule `no-loop-func` is literally the rule which disallows all function declarations in loop. The purpose of the rule is avoidance of confusion by the fact that loop-blocks don't make any scope for variables.

```js
for (var i = 0; i < 3; ++i) {
    setTimeout(() => console.log(i)) // 3, 3, 3
}
```

> [Online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiZm9yICh2YXIgaSA9IDA7IGkgPCAzOyArK2kpIHtcbiAgICBzZXRUaW1lb3V0KCgpID0+IGNvbnNvbGUubG9nKGkpKSAvLyAzLCAzLCAzXG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo2LCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7ImltcGxpZWRTdHJpY3QiOnRydWV9fSwicnVsZXMiOnsibm8tbG9vcC1mdW5jIjoyfSwiZW52Ijp7InNoYXJlZC1ub2RlLWJyb3dzZXIiOnRydWV9fX0=)

Since ES2015, loop-blocks can make the scope of `let` and `const` variables. So the rule got some relaxing. The rule allows functions which don't use variables of outer scopes than the loop.

```js
for (let i = 0; i < 3; ++i) {
    setTimeout(() => console.log(i)) // 1,2,3; OK because the `i` is in the loop-block's scope.
}

let outer
for (let i = 0; i < 3; ++i) {
    setTimeout(() => console.log(outer)) // 3,3,3; Disallowed because the `outer` is not in the loop-block's scope.
    outer += 1
}

for (let i = 0; i < 3; ++i) {
    setTimeout(() => console.log(outer)) // 3,3,3; OK because the `outer` is not modified after this loop.
}
```

> [Online demo](https://eslint.org/demo/#eyJ0ZXh0IjoiZm9yIChsZXQgaSA9IDA7IGkgPCAzOyArK2kpIHtcbiAgICBzZXRUaW1lb3V0KCgpID0+IGNvbnNvbGUubG9nKGkpKSAvLyAxLDIsMzsgT0sgYmVjYXVzZSB0aGUgYGlgIGlzIGluIHRoZSBsb29wLWJsb2NrJ3Mgc2NvcGUuXG59XG5cbmxldCBvdXRlclxuZm9yIChsZXQgaSA9IDA7IGkgPCAzOyArK2kpIHtcbiAgICBzZXRUaW1lb3V0KCgpID0+IGNvbnNvbGUubG9nKG91dGVyKSkgLy8gMywzLDM7IERpc2FsbG93ZWQgYmVjYXVzZSB0aGUgYG91dGVyYCBpcyBub3QgaW4gdGhlIGxvb3AtYmxvY2sncyBzY29wZS5cbiAgICBvdXRlciArPSAxXG59XG5cbmZvciAobGV0IGkgPSAwOyBpIDwgMzsgKytpKSB7XG4gICAgc2V0VGltZW91dCgoKSA9PiBjb25zb2xlLmxvZyhvdXRlcikpIC8vIDMsMywzOyBPSyBiZWNhdXNlIHRoZSBgb3V0ZXJgIGlzIG5vdCBtb2RpZmllZCBhZnRlciB0aGlzIGxvb3AuXG59XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnsiaW1wbGllZFN0cmljdCI6dHJ1ZX19LCJydWxlcyI6eyJuby1sb29wLWZ1bmMiOjJ9LCJlbnYiOnsic2hhcmVkLW5vZGUtYnJvd3NlciI6dHJ1ZX19fQ==)

In your example, the function is using variables which are in outer scopes than the loop.

**Comment by isiahmeadows** [comment]
Oh okay. So I guess I'll keep my workaround of a separate function outside the loop. Could the last two cases and how they differ (where `outer` is unmodified outside the loop) be clarified in the docs, then?

**Comment by mayankraipure** [comment]
Use this
/*eslint no-loop-func: "off"*/
Instead of
/*eslint no-loop-func: "error"*/

**Comment by rachx** [comment]
Working on this
